File: codejam/cpp/2010/1B/1B_YourRankIsPure.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pi;
#define FOR(i,a) for (ll i = 0; i < (a); i++)
#define len(x) (ll) x.size()
const ll INF = 1LL << 62;
const ll MOD = 100003;
//const ll MOD = 998244353;
const double PI = 4*atan(double(1.0));
vector<vector<ll>> twodi(ll N, ll M, ll v) { vector<vector<ll>> res(N); for (auto &vv : res) vv.resize(M,v); return res; }
ll dp[501][501];
ll comb[501][501];
ll ansarr[501];

int main (int argc, char **argv) {
    if (argc > 1) { freopen(argv[1],"r",stdin); }
    ios_base::sync_with_stdio(false);cin.tie(0);
    // PROGRAM STARTS HERE
    // BEGIN PREWORK
    FOR(i,501) FOR(j,i+1) comb[i][j] = (j==0||j==i) ? 1 : (comb[i-1][j-1]+comb[i-1][j]) % MOD;
    FOR(i,501) FOR(j,501) dp[i][j] = 0;
    for (ll i=2; i<=500; i++) {
        dp[i][1] = 1;
        for (ll k=2;k<i;k++) {
            for (ll l=1;l<k;l++) dp[i][k] += (comb[i-k-1][k-l-1] * dp[k][l]) % MOD;
            dp[i][k] %= MOD;
        }
    }
    for (ll i=0;i<=500;i++) {
        ansarr[i] = 0;
        for (ll j=0;j<=500;j++) ansarr[i] += dp[i][j];
        ansarr[i] %= MOD;
    }
    // END PREWORK
    ll T; cin >> T;
    for (ll tt=1;tt<=T;tt++) {
        ll N; cin >> N;
        auto ans = ansarr[N];
        printf("Case #%lld: %lld\n",tt,ans);
    }
}

File: codejam/python/2010/1B/1B_YourRankIsPure.py
import sys
from collections import deque

## Input crap
infile = sys.stdin
intokens = deque()
def getTokens(): 
    while not intokens:
        for c in infile.readline().rstrip().split() :
            intokens.append(c)    
def gs(): getTokens(); return intokens.popleft()
def gi(): return int(gs())
def gf(): return float(gs())
def gbs(): return [c for c in gs()]
def gis(n): return [gi() for i in range(n)]
def ia(m): return [0] * m
def iai(m,v): return [v] * m
def twodi(n,m,v): return [iai(m,v) for i in range(n)]
def fill2(m) : r = gis(2*m); return r[0::2],r[1::2]
def fill3(m) : r = gis(3*m); return r[0::3],r[1::3],r[2::3]
def fill4(m) : r = gis(4*m); return r[0::4],r[1::4],r[2::4],r[3::4]
MOD = 100003
##MOD = 1000000007

def main() :
    if len(sys.argv) > 1 : global infile; infile = open(sys.argv[1],'rt')
    ## PROGRAM STARTS HERE
    T = gi()

    ## BEGIN PREWORK
    comb = [[0]*501 for _ in range(501)]
    for i in range(501) :
        for j in range(i+1) :
            comb[i][j] = 1 if j == 0 or j == i else (comb[i-1][j-1]+comb[i-1][j]) % MOD
    dp = [[0]*501 for _ in range(501)]
    for i in range(2,501) :
        for k in range(1,i) :
            if k == 1 : dp[i][k] = 1; continue
            for l in range(1,k) :
                dp[i][k] += comb[i-k-1][k-l-1] * dp[k][l]
    ansarr = [ sum(dp[i]) % MOD for i in range(501) ]
    ## END PREWORK

    for tt in range(1,T+1) :
        n = gi(); ans = ansarr[n]; print(f"Case #{tt}: {ans}")

if __name__ == "__main__" :
    main()

